README.md
lazBackup

An attemp at an Archive / backup programme.


history.txt
Build 1

Initial conversion from stub.

Currently
1174 lines of source code.
552 statements
17.4% 


Build 2

The display now consists of three list boxes, one for the source directory,
one for the destination director and the middle one for the differences.

Added logic to walk the source and destination directories, after both have
been specified.

Added the Analyse function which determined what files have been added to the
source and not in destination, have been deleted from the destination i.e. no
longer in source.

Currently
1430 lines of source code.
701 statements
17.6% 


formhelp.pas
unit formhelp;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, FileUtil, Forms, Controls, Graphics, Dialogs, StdCtrls,
  ExtCtrls;

type

  { TfrmHelp }

  TfrmHelp = class(TForm)
    btnhelpExit : TButton;
    lblComments : TLabel;
    lblCopyRight: TLabel;
    lblVersion  : TLabel;
    mmoHelp     : TMemo;
    Panel1      : TPanel;
    Panel2      : TPanel;

    procedure btnhelpExitClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    { private declarations }
  public
    { public declarations }
  end; 

var
  frmHelp : TfrmHelp;

implementation

uses
  formBackup;

{$R *.lfm}

{ TfrmHelp }

procedure TfrmHelp.FormCreate(Sender: TObject);
begin
  mmoHelp.Append('Stub.');
  mmoHelp.Append('');
  mmoHelp.Append('');

  try
    mmoHelp.Lines.LoadFromFile('help.txt');
  except
    on Exception do begin
      mmoHelp.Append(' help file not found.');
      mmoHelp.Append('');
      mmoHelp.Append(' This file should include full and detailed help intructions.');
    end;
  end;

  lblComments.Caption  := format('%s :: %s', [userOptions.productName,userOptions.Comments]);
  lblCopyRight.Caption := userOptions.legalCopyright;
  lblVersion.Caption   := format('%s Version :: %s', [userOptions.productName, userOptions.fileVersion]);
end;

procedure TfrmHelp.btnhelpExitClick(Sender: TObject);
begin
  Close;
end;


end.



formabout.pas
unit formAbout;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, FileUtil, Forms, Controls, Graphics, Dialogs, StdCtrls,
  LCLVersion, ExtCtrls, ShellApi, backupUtils;

type

  { TfrmAbout }

  TfrmAbout = class(TForm)
    btnAboutExit         : TButton;
    btnAboutMSInfo       : TButton;
    Image1               : TImage;
    lblProgramName       : TLabel;
    lblProgrammer        : TLabel;
    lblSysUpTime         : TLabel;
    lblAppUpTime         : TLabel;
    lblSystemUpTime      : TLabel;
    lblApplicationUpTime : TLabel;
    LstBxInfo            : TListBox;
    LstBxDiscSpace       : TListBox;
    Panel1               : TPanel;
    Panel2               : TPanel;
    Panel3               : TPanel;
    Panel4               : TPanel;
    tmrUpTime            : TTimer;

    procedure btnAboutExitClick(Sender: TObject);
    procedure btnAboutMSInfoClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure tmrUpTimeTimer(Sender: TObject);
  private
    { private declarations }
  public
    { public declarations }
  end; 

var
  frmAbout: TfrmAbout;

implementation

uses
  formBackup;

{$R *.lfm}

{ TfrmAbout }


procedure TfrmAbout.btnAboutExitClick(Sender: TObject);
{  Close About form.  }
begin
  tmrUpTime.Enabled := false;
  Close;
end;

procedure TfrmAbout.btnAboutMSinfoClick(Sender: TObject);
{  Run the external application MSinfo.
   If ShellExecute returns 32 or less this indicates an error, just inform the user.
}
begin
  if ShellExecute(0, nil, PChar('"msinfo32.exe"'), nil, nil, 1) < 33 then
    ShowMessage('ERROR : running MSinfo');
end;

procedure TfrmAbout.FormCreate(Sender: TObject);
var
  dskSize: string;
  dskFree: string;
  message: string;
  i      : integer;
begin
  tmrUpTime.Enabled     := True;
  lblAppUpTime.Caption  := getUpTime('Application');
  lblSysUpTime.Caption  := getUpTime('System');
  lblProgrammer.Caption := userOptions.legalCopyright;

  lstBxInfo.Items.add(userOptions.fileDescription);
  lstBxInfo.Items.add('');
  lstBxInfo.Items.add(format('lazKlock Build   :: %s', [userOptions.productVersion]));
  lstBxInfo.Items.add(format('lazKlock Version :: %s', [userOptions.fileVersion]));
  lstBxInfo.Items.add(format('lazKlock Built :: %s', [FormatDateTime('DD/MMM/YYYY hh:nn:ss : ', now)]));
  {$ifdef WIN32}
    lstBxInfo.Items.add(format('Built with 32 bit Lazarus Version :: %s', [lcl_version]));
  {$else}
    lstBxInfo.Items.add(format('Built with 64 bit Lazarus Version :: %s', [lcl_version]));
  {$endif}
  lstBxInfo.Items.add('');
  lstBxInfo.Items.add(getWindowsVersion);
  lstBxInfo.Items.add('');
  lstBxInfo.Items.add(userOptions.CompanyName);
  lstBxInfo.Items.add('');
  lstBxInfo.Items.add(userOptions.Comments);
  lstBxInfo.Items.add('');
  lstBxInfo.Items.add('App Dir : ' + ExtractFilePath(Application.ExeName));

  // Display the free space on drives B, C, D, E, F, where present
  for i := 2 to 10 do
  begin
    dskFree := FloatToStrF(DiskFree(i) / 1073741824, ffFixed, 3, 2);
    dskSize := FloatToStrF(DiskSize(i) / 1073741824, ffFixed, 3, 2);

    if DiskSize(i) >= 0 then
    begin
      message := format(' Disk %s : Free / Size :: %s / %s Gbytes', [Chr(i + 64), dskFree, dskSize]);
      LstBxDiscSpace.Items.Add(message);
    end;
  end;

end;

procedure TfrmAbout.tmrUpTimeTimer(Sender: TObject);
{  Update the labels in real time.    }
begin
  lblAppUpTime.Caption := getUpTime('Application');
  lblSysUpTime.Caption := getUpTime('System');
end;

end.



backuputils.pas
unit backupUtils;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, strutils, process, Graphics;

function FontToString(f: TFont): string;
function StringToFont(s: string): TFont;
function getUpTime(system: string): string;
function getWindowsVersion: string;

implementation

uses
  formBackup;

function FontToString(f: TFont): string;
{  Produces a string representation of a given font.
   All the font attributes are converted to strings and packed together.

   Does not yet handle errors - will there be any ;o)                                              }
var
  rtnStr: string;
begin
  rtnStr := '';
  rtnStr := rtnStr + IntToStr(f.Charset) + ':';
  rtnStr := rtnStr + ColorToString(f.Color) + ':';
  rtnStr := rtnStr + IntToStr(f.Height) + ':';
  rtnStr := rtnStr + (f.Name) + ':';
  rtnStr := rtnStr + IntToStr(f.Orientation) + ':';
  rtnStr := rtnStr + IntToStr(f.size) + ':';
  rtnStr := rtnStr + IntToStr(Ord(f.Pitch)) + ':';
  rtnStr := rtnStr + IntToStr(Ord(f.Quality)) + ':';

  //                             TFontStyles is a set and scanned differently.
  if fsBold in f.Style then
    rtnStr := rtnStr + 'B'
  else
    rtnStr := rtnStr + '.';

  if fsItalic in f.Style then
    rtnStr := rtnStr + 'I'
  else
    rtnStr := rtnStr + '.';

  if fsStrikeOut in f.Style then
    rtnStr := rtnStr + 'S'
  else
    rtnStr := rtnStr + '.';

  if fsUnderline in f.Style then
    rtnStr := rtnStr + 'U'
  else
    rtnStr := rtnStr + '.';


  Result := rtnStr;
end;

function StringToFont(s: string): TFont;
{  Produces a font from a given string representation [produced by FonttoString.
   The string is read a bit at at time [up to the next :] and this is converted
   to the given font attribute.

   NB :: pos starts from position 1, where copy start at position 1.
         TFontStyles is a set and scanned differently.

   Does not yet handle errors - will there be any ;o)                                              }
var
  p   : integer;
  err : integer;
  chrs: integer;
  clr : TColor;
  Hght: integer;
  nme : string;
  Ortn: integer;
  sze : integer;
  ptch: integer;
  qlty: integer;

  fnt : TFont;
  fstyles: TFontStyles;
begin
  fnt := TFont.Create;
  fstyles := [];             //  empty set ??

  p := Pos(':', s);                                   //  Character set of font
  val(copy(s, 0, p - 1), chrs, err);
  //if err = 0 then klog.writeLog(format('StringToFont error : %d', [err]));
  Delete(s, 1, p);

  p := Pos(':', s);                                   //  colour of font
  clr := StringToColor(copy(s, 0, p - 1));
  Delete(s, 1, p);

  p := Pos(':', s);                                  //  height of font
  val(copy(s, 0, p - 1), Hght, err);
  //if err = 0 then klog.writeLog(format('StringToFont error : %d', [err]));
  Delete(s, 1, p);

  p := Pos(':', s);                                  //  name of font
  nme := copy(s, 0, p - 1);
  Delete(s, 1, p);

  p := Pos(':', s);                                  //  orientation of font
  val(copy(s, 0, p - 1), Ortn, err);
  //if err = 0 then klog.writeLog(format('StringToFont error : %d', [err]));
  Delete(s, 1, p);

  p := Pos(':', s);                                  //  size of font
  val(copy(s, 0, p - 1), sze, err);
  //if err = 0 then klog.writeLog(format('StringToFont error : %d', [err]));
  Delete(s, 1, p);

  p := Pos(':', s);                                  //  pitch of font
  val(copy(s, 0, p - 1), ptch, err);
  //if err = 0 then klog.writeLog(format('StringToFont error : %d', [err]));
  Delete(s, 1, p);

  p := Pos(':', s);                                 //  quality of font
  val(copy(s, 0, p - 1), qlty, err);
  //if err = 0 then klog.writeLog(format('StringToFont error : %d', [err]));
  Delete(s, 1, p);

  if Pos('B', s) <> 0 then
    include(fstyles, fsBold);
  if Pos('I', s) <> 0 then
    include(fstyles, fsItalic);
  if Pos('S', s) <> 0 then
    include(fstyles, fsStrikeOut);
  if Pos('U', s) <> 0 then
    include(fstyles, fsUnderline);

  fnt.Charset     := chrs;                 //  Character set of font
  fnt.Color       := clr;                  //  colour of font
  fnt.Height      := Hght;                 //  height of font
  fnt.Name        := nme;                  //  name of font
  fnt.Orientation := Ortn;                 //  orientation of font
  fnt.Size        := sze;                  //  size of font
  fnt.Pitch       := TFontPitch(ptch);     //  pitch of font
  fnt.Quality     := TFontQuality(qlty);   //  quality of font
  fnt.Style       := fstyles;

  Result := fnt;
end;

function getUpTime(system: string): string;
{  Determines the up time of either System or Application - depending on argument S or A.
   Used in formAbout & uKlockUtils.

   appStartTime := GetTickCount64; needs to be run when the app starts.

   NOTE :: Windows Only - use LclIntf.GetTickCount for cross platform.

   TODO : Need to check for roll over and account for it.
}
var
  noTicks    : int64;
  noSeconds  : integer;
  noOfDays   : integer;
  noOfHours  : integer;
  noOfMinutes: integer;
  noOfSeconds: integer;
begin
  system := AnsiLowerCase(system);

  if AnsiStartsStr('s', system) then
    noTicks := GetTickCount64                     //  How long has the system been running.
  else
    noTicks := GetTickCount64 - appStartTime;     //  How long has the application been running.

  noSeconds := noTicks div 1000;                  //  1000 ticks per second.

  noOfDays    := noSeconds div 86400;
  noSeconds   := noSeconds - (noOfDays * 86400);
  noOfHours   := noSeconds div 3600;
  noSeconds   := noSeconds - (noOfHours * 3600);
  noOfMinutes := noSeconds div 60;
  noSeconds   := noSeconds - (noOfMinutes * 60);
  NoOfSeconds := noSeconds;

  Result := format('%d days : %d hours : %d mins : %d secs', [noOfDays, noOfHours, noOfMinutes, noOfSeconds]);
end;

function getWindowsVersion: string;
{  Gets the version of the windows OS.
   This is achieved by capturing the output from the DOS command ver.
   The ver command is run  into a file and then the file is read back.
}
VAR
  AProcess   : TProcess;
  winVer     : TStringList;
  tmpWinVer  : string;
  tmpFileName: string;
begin
  tmpFileName := GetTempDir(true) + 'ver.txt';

  AProcess            := TProcess.Create(nil);
  AProcess.Options    := [poWaitOnExit, poUsePipes];
  AProcess.Executable := 'CMD ';
  AProcess.Parameters.Add('/C ver >' + tmpFileName);
  AProcess.Execute;

  winVer := TStringList.Create;
  winVer.LoadFromFile(tmpFileName);
  tmpWinVer := winVer[1];

  AProcess.Free;
  winVer.Free;

  Result := tmpWinVer;
end;


end.



formlicence.pas
unit formLicence;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, FileUtil, Forms, Controls, Graphics, Dialogs, StdCtrls,
  ExtCtrls;

type

  { TfrmLicence }

  TfrmLicence = class(TForm)
    btnLicenceExit: TButton;
    lblComments   : TLabel;
    lblCopyRight  : TLabel;
    lblVersion    : TLabel;
    mmoLicence    : TMemo;
    Panel1        : TPanel;
    Panel2        : TPanel;

    procedure btnLicenceExitClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    { private declarations }
  public
    { public declarations }
  end; 

var
  frmLicence: TfrmLicence;

implementation

uses
  formBackup;

{$R *.lfm}

{ TfrmLicence }

procedure TfrmLicence.FormCreate(Sender: TObject);
begin
  mmoLicence.Append('Stub.');
  mmoLicence.Append('');
  try
    mmoLicence.Lines.LoadFromFile('GNU GENERAL PUBLIC LICENSE.txt');
  except
    on Exception do begin
      mmoLicence.Append(' help License not found.');
      mmoLicence.Append('');
      mmoLicence.Append(' The application is issued under the GNU GENERAL PUBLIC LICENSE.');
    end;
  end;

  lblComments.Caption  := userOptions.Comments;
  lblCopyRight.Caption := userOptions.legalCopyright;
  lblVersion.Caption   := format('%s Version :: %s', [userOptions.productName, userOptions.fileVersion]);
end;

procedure TfrmLicence.btnLicenceExitClick(Sender: TObject);
begin
  Close;
end;

end.



uoptions.pas
unit uOptions;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, DOM, XMLWrite, XMLRead, fileinfo, winpeimagereader, Dialogs,
  Graphics, backupUtils;

type
  Options = class
  {  Holds the current user options.  The options are hard coded in this class.

     When the class is created, the constructor first checks that the user directory
     exists - this should be something like c:\Users\<user>\AppData\Local\<app Name>\Options.xml.
     The file name can be passed to the contructor, if it's absent the 'Options.xml' is used.
     If this directory does not exist, it will be created.

     Next, the user options file is checked.  If it is exists, it is read.
     If the file does not exist - the file will be created with default values.

     class methods -
         create                       - creates the options class with a default filename.
         create(fileName: string)     - creates the options class with a specified filename.
         readOptions                  - reads the options file and populates the options class
         writeCurrentOptions          - writes the current options to the options filename [in XML].
         writeDefaultOptions          - writes the options file with default files [internal use really]
                                      - could be used to reset options.

     The writeCurrentOptions should be executed when the program closes, or when options change.
     That is if the changes need to be saved.

     NOTE :: If there is an error while either reading or writing the options file, the application is halted.

     NOTE :: All XML values are string, so need to be casted before use - this is done in the read / write routines.

  }
    private
      _dirName: String;
      //  Global
      _Comments        : string;
      _companyName     : string;
      _fileDescription : string;
      _fileVersion     : string;
      _InternalName    : string;
      _legalCopyright  : string;
      _originalFileName: string;
      _productName     : string;
      _productVersion  : string;

      _screenSave: boolean;           //  do we save Klock position or not.
      _formTop   : integer;           //  the forms top left.
      _formLeft  : integer;

      _optionsName: string;           //  full path to the options file.

      procedure checkDirectory;
      Function readChild(         PassNode: TDOMNode;  name: string): string;
      Function readChildAttribute(PassNode: TDOMNode;  name: string; attribute: string): string;

      function writeStrChild(Doc: TXMLDocument; name: string; value: string): TDOMNode;
      function writeFontChild(Doc: TXMLDocument; name: string; value: Tfont): TDOMNode;
      function writeColChild(Doc: TXMLDocument; name: string; value: TColor): TDOMNode;
      function writeBolChild(Doc: TXMLDocument; name: string; value: boolean): TDOMNode;
      function writeIntChild(Doc: TXMLDocument; name: string; value: integer): TDOMNode;
      function writeFloatChild(Doc: TXMLDocument; name: string; value: Double): TDOMNode;
      function writeIntChildAttribute(Doc: TXMLDocument; name: string; value1: integer; value2: integer): TDOMNode;
    public
      //  Global - file stuff
      property Comments        : String read _Comments         write _Comments;
      property companyName     : String read _companyName      write _companyName;
      property fileDescription : String read _fileDescription  write _fileDescription;
      property fileVersion     : String read _fileVersion      write _fileVersion;
      property InternalName    : String read _InternalName     write _InternalName;
      property legalCopyright  : String read _legalCopyright   write _legalCopyright;
      property originalFileName: String read _originalFileName write _originalFileName;
      property productName     : String read _productName      write _productName;
      property productVersion  : String read _productVersion   write _productVersion;

      //  Global - other stuff
      property screenSave : boolean read _screenSave write _screenSave;
      property formTop    : integer read _formTop    write _formTop;
      property formLeft   : integer read _formLeft   write _formLeft;
      property optionsName: string read _optionsName write _optionsName;

      constructor Create; overload;
      constructor Create(filename: String); overload;

      procedure readOptions;
      procedure Assign(o: Options);
      procedure writeCurrentOptions;
      procedure writeDefaultOptions;
  end;  //  class

  myFileVersionInfo = class
  {  Retrieves the current file info.

     see http://wiki.freepascal.org/Show_Application_Title,_Version,_and_Company.

     NOTE :: The information in Project / Project Options / Version Ino in the IDE
     must be up to date.
  }

    private
      _comments        : String;
      _companyName     : String;
      _fileDescription : String;
      _fileVersion     : String;
      _InternalName    : String;
      _legalCopyright  : String;
      _originalFileName: String;
      _productName     : String;
      _productVersion  : String;

    public
      property fileComments        : String read _comments         write _comments;
      property fileCompanyName     : String read _companyName      write _companyName;
      property fileFileDescription : String read _fileDescription  write _fileDescription;
      property fileFileVersion     : String read _fileVersion      write _fileVersion;
      property fileInternalName    : String read _InternalName     write _InternalName;
      property fileLegalCopyright  : String read _legalCopyright   write _legalCopyright;
      property fileOriginalFileName: String read _originalFileName write _originalFileName;
      property fileProductName     : String read _productName      write _productName;
      property fileProductVersion  : String read _productVersion   write _productVersion;

      procedure GetFileInfo;
  end;

implementation

//
//............................................ Options methods ............................................
//

  constructor Options.Create; overload;
  {  creates the options class with a default filename.  }
  var
  optnFile: String;
  begin
    checkDirectory;

    {$IFDEF TEST}
      optnFile := 'TEST_Options';
    {$else}
      optnFile := 'Options';
    {$endif}
    {$ifdef WIN32}
      optionsName := _dirName + optnFile + '32.xml';
    {$else}
      optionsName := _dirName + optnFile + '64.xml';
    {$endif}

    If FileExists(optionsName) Then
    begin
      writeDefaultOptions;       //  Set up the defaults [which includes any new values]
      readOptions;               //  read the exiting options file.
      writeCurrentOptions;       //  write out a new options file [which includes any new values]
    end
    else
      writeDefaultOptions;       //  options file not found, write out a new one.
  end;

  constructor Options.Create(fileName: String); overload;
  {  creates the options class with a specified filename.  }
  begin
    checkDirectory;

    optionsName := _dirName + fileName;

    If FileExists(optionsName) Then
    begin
      writeDefaultOptions;       //  Set up the defaults [which includes any new values]
      readOptions;               //  read the exiting options file.
      writeCurrentOptions;       //  write out a new options file [which includes any new values]
    end
    else
      writeDefaultOptions;       //  options file not found, write out a new one.
  end;

  procedure Options.checkDirectory;
  {  Checks that the options directory exists.

     GetAppConfigDir(False) -> c:\Users\<user>\AppData\Local\<app Name>\
     GetAppConfigDir(True)  -> c:\ProgramData\<app Name>\
  }
  begin
    _dirName := GetAppConfigDir(False);

    If NOT DirectoryExists(_dirName) Then
      If Not CreateDir (_dirName) Then
        ShowMessage('Failed to create directory !');
  end;

  procedure Options.Assign(o: Options);
  {  Copy all fields from one options object to another.
   Because Options is derived from TObjects and not TPersistent, we don't get assign for free.

   NOTE :: When a new field of added to the Option class, it HAS to be added here.
           Must be a better way of doing this.
  }
  begin
  //  Global - file stuff
  Comments         := o.Comments;
  companyName      := o.companyName;
  fileDescription  := o.fileDescription;
  fileVersion      := o.fileVersion;
  InternalName     := o.InternalName;
  legalCopyright   := o.legalCopyright;
  originalFileName := o.originalFileName;
  productName      := o.productName;
  productVersion   := o.productVersion;

  //  Global - other stuff
  optionsName := o.optionsName;
  screenSave  := o.screenSave;
  formTop     := o.formTop;
  formLeft    := o.formLeft;
  end;

  procedure Options.readOptions;
  {  Read in the options file.
   The filename is specified when the user options class is created.
   The file info is re-read, in case is has changed.

   The read is now a two stage process.
   Stage 1 - the xml node is read, this return a string value.
   stage 2 - if the return is 'ERROR', the xml node is missing and then use the
             default value.  If not the return will hold the value which is passed
             to the options property.

   NOTE : This cures the missing child problem, BUT NOT the missing node.
  }
  var
    fvi      : myFileVersionInfo;
    PassNode : TDOMNode;
    Doc      : TXMLDocument;
    rtn      : string;
  begin
    try
      //  retrieve file info i.e build numner etc.
      fvi:= myFileVersionInfo.create;
      fvi.GetFileInfo;

      Comments         := fvi.fileComments;
      companyName      := fvi.fileCompanyName;
      fileDescription  := fvi.fileFileDescription;
      fileVersion      := fvi.fileFileVersion;
      InternalName     := fvi.fileInternalName;
      legalCopyright   := fvi.fileLegalCopyright;
      originalFileName := fvi.fileOriginalFileName;
      productName      := fvi.fileProductName;
      productVersion   := fvi.fileProductVersion;

      try
        // Read in xml file from disk
        ReadXMLFile(Doc, optionsName);
      except
        on E: Exception do
        begin
          ShowMessage('ERROR: reading XML file.' + LineEnding
                    + E.Message + LineEnding
                    + 'Halting Program Execution');
          Halt;
        end;  //  on E:
      end;    //  try

      //  Global
      PassNode := Doc.DocumentElement.FindNode('Global');

      if assigned(PassNode) then
      begin
        rtn := readChild(PassNode, 'screenSave');
        if rtn <> 'ERROR' then screenSave := StrToBool(rtn);
        rtn := readChildAttribute(PassNode, 'formPosition', 'Top');
        if rtn <> 'ERROR' then formTop := StrToInt(rtn);
        rtn := readChildAttribute(PassNode, 'formPosition', 'Left');
        if rtn <> 'ERROR' then formLeft := StrToInt(rtn);

        rtn := readChild(PassNode, 'optionsName');
        if rtn <> 'ERROR' then optionsName := ansistring(rtn);
      end;

    finally
      // finally, free the document
      Doc.Free;
    end;
  end;

  procedure Options.writeDefaultOptions;
  {  Sets us some sensible defaults and then calls writeCurrentOptions to writs out the xml file.
     Used if the useroptions file does not exist.
     The filename is specified when the user options class is created.
     The file info is re-read, in case is has changed
  }
  var
  fvi: myFileVersionInfo;
  begin
    //  retrieve file info i.e build numner etc.
    fvi:= myFileVersionInfo.create;
    fvi.GetFileInfo;

    Comments         := fvi.fileComments ;
    companyName      := fvi.fileCompanyName;
    fileDescription  := fvi.fileFileDescription;
    fileVersion      := fvi.fileFileVersion;
    InternalName     := fvi.fileInternalName;
    legalCopyright   := fvi.fileLegalCopyright;
    originalFileName := fvi.fileOriginalFileName;
    productName      := fvi.fileProductName;
    productVersion   := fvi.fileProductVersion;

    screenSave := True;
    formTop    := 100;              //  the forms top left.
    formLeft   := 100;

    writeCurrentOptions;
  end;

  procedure Options.writeCurrentOptions;
  {  Writes out the user options to a xml file.
     The filename is specified when the user options class is created.
     The file info is re-read, in case is has changed
  }
  var
  Doc        : TXMLDocument;
  RootNode   : TDOMNode;
  ElementNode: TDOMNode;
  fvi        : myFileVersionInfo;
  begin
    try
      //  retrieve file info i.e build numner etc.
      fvi:= myFileVersionInfo.create;
      fvi.GetFileInfo;

      // Create a document
      Doc := TXMLDocument.Create;

      // Create a root node
      RootNode := Doc.CreateElement('Klock');
      Doc.Appendchild(RootNode);
      RootNode:= Doc.DocumentElement;

      //  Global
      ElementNode:=Doc.CreateElement('Global');

      ElementNode.AppendChild(writeStrChild(doc, 'Comments',         fvi.fileComments));
      ElementNode.AppendChild(writeStrChild(doc, 'companyName',      fvi.filecompanyName));
      ElementNode.AppendChild(writeStrChild(doc, 'fileDescription',  fvi.filefileDescription));
      ElementNode.AppendChild(writeStrChild(doc, 'fileVersion',      fvi.filefileVersion));
      ElementNode.AppendChild(writeStrChild(doc, 'InternalName',     fvi.fileInternalName));
      ElementNode.AppendChild(writeStrChild(doc, 'legalCopyright',   fvi.fileLegalCopyright));
      ElementNode.AppendChild(writeStrChild(doc, 'originalFileName', fvi.fileOriginalFileName));
      ElementNode.AppendChild(writeStrChild(doc, 'productName',      fvi.fileProductName));
      ElementNode.AppendChild(writeStrChild(doc, 'productVersion',   fvi.fileProductVersion));

      ElementNode.AppendChild(writeStrChild(doc, 'optionsName', optionsName));

      ElementNode.AppendChild(writeBolChild(doc, 'screenSave', screenSave));
      ElementNode.AppendChild(writeIntChildAttribute(Doc, 'formPosition', formTop, formLeft));

      try
        // Save XML
        WriteXMLFile(Doc, optionsName);
      except
        on E: Exception do
        begin
          ShowMessage('ERROR: Writing XML file.' + LineEnding
                    + E.Message + LineEnding
                    + 'Halting Program Execution');
          Halt;
        end;  //  on E:
      end;    //  try

    finally
      Doc.Free;
    end;
  end;

//
//........................................ fileVersionInfo methods ............................................
//

  procedure myFileVersionInfo.GetFileInfo;
  {  Retrieves the file info from the current file.
     Called from the class myFileVersionInfo.
  }
  var
    FileVerInfo: TFileVersionInfo;
  begin
    FileVerInfo:=TFileVersionInfo.Create(nil);

    try
      FileVerInfo.ReadFileInfo;

      fileComments         := FileVerInfo.VersionStrings.Values['Comments'];
      fileCompanyName      := FileVerInfo.VersionStrings.Values['CompanyName'];
      fileFileDescription  := FileVerInfo.VersionStrings.Values['FileDescription'];
      fileFileVersion      := FileVerInfo.VersionStrings.Values['FileVersion'];
      fileInternalName     := FileVerInfo.VersionStrings.Values['InternalName'];
      fileLegalCopyright   := FileVerInfo.VersionStrings.Values['LegalCopyright'];
      fileOriginalFileName := FileVerInfo.VersionStrings.Values['OriginalFilename'];
      fileProductName      := FileVerInfo.VersionStrings.Values['ProductName'];
      fileProductVersion   := FileVerInfo.VersionStrings.Values['ProductVersion'];
    finally
          FileVerInfo.Free;
    end;

  end;
   //
  //........................................ Helper functions ....................
  //
  Function Options.readChild(PassNode: TDOMNode;  name: string): string;
  {  Read a child node and return its [string] value.    }
  var
      childNode: TDOMNode;
  begin
      childNode := PassNode.FindNode(name);

      if assigned(childNode) then
        result := childNode.TextContent
      else
        result := 'ERROR';
  end;

  Function Options.readChildAttribute(PassNode: TDOMNode;  name: string; attribute: string): string;
  {  Read a child node and return its named [string] attribute.    }
  var
      childNode: TDOMNode;
  begin
    childNode := PassNode.FindNode(name);

    if assigned(childNode) then
      result := TDOMElement(childNode).GetAttribute(attribute)
    else
      result := 'ERROR';
  end;

  function Options.writeStrChild(Doc: TXMLDocument; name: string; value: string): TDOMNode;
  {  Write a [string] value to a child node.    }
  var
    ItemNode, TextNode: TDOMNode;
  begin
    ItemNode := Doc.CreateElement(name);
    TextNode := Doc.CreateTextNode(WideString(value));
    ItemNode.AppendChild(TextNode);
    result := ItemNode;
  end;

  function Options.writeColChild(Doc: TXMLDocument; name: string; value: TColor): TDOMNode;
  {  Write a [string] value to a child node.    }
  var
    ItemNode, TextNode: TDOMNode;
  begin
    ItemNode := Doc.CreateElement(name);
    TextNode := Doc.CreateTextNode(ColorToString(value));
    ItemNode.AppendChild(TextNode);
    result := ItemNode;
  end;

  function Options.writeFontChild(Doc: TXMLDocument; name: string; value: TFont): TDOMNode;
  {  Write a [font] value to a child node.    }
  var
    ItemNode, TextNode: TDOMNode;
  begin
    ItemNode := Doc.CreateElement(name);
    TextNode := Doc.CreateTextNode(FontToString(value));
    ItemNode.AppendChild(TextNode);
    result := ItemNode;
  end;

  function Options.writeBolChild(Doc: TXMLDocument; name: string; value: boolean): TDOMNode;
  {  Write a [boolean] value to a child node.    }
  var
    ItemNode, TextNode: TDOMNode;
  begin
    ItemNode := Doc.CreateElement(name);
    TextNode := Doc.CreateTextNode(BoolToStr(value));
    ItemNode.AppendChild(TextNode);
    result := ItemNode;
  end;

  function Options.writeFloatChild(Doc: TXMLDocument; name: string; value: Double): TDOMNode;
  {  Write a [boolean] value to a child node.    }
  var
    ItemNode, TextNode: TDOMNode;
  begin
    ItemNode := Doc.CreateElement(name);
    TextNode := Doc.CreateTextNode(FloatToStr(value));
    ItemNode.AppendChild(TextNode);
    result := ItemNode;
  end;

  function Options.writeIntChild(Doc: TXMLDocument; name: string; value: integer): TDOMNode;
  {  Write a [integer] value to a child node.    }
  var
    ItemNode, TextNode: TDOMNode;
  begin
    ItemNode := Doc.CreateElement(name);
    TextNode := Doc.CreateTextNode(IntToStr(value));
    ItemNode.AppendChild(TextNode);
    result := ItemNode;
  end;

  function Options.writeIntChildAttribute(Doc: TXMLDocument; name: string; value1: integer; value2: integer): TDOMNode;
  {  Write a [integer] attribute to a child node.

     It seems you have to write both attributes at once, i can read them singularly.
     So, this routine is hard coded for form position until i can fix.
  }
  var
    ItemNode, TextNode: TDOMNode;
  begin
    ItemNode := Doc.CreateElement(name);
    TDOMElement(ItemNode).SetAttribute('Top', IntToStr(value1));
    TDOMElement(ItemNode).SetAttribute('Left', IntToStr(value2));
    TextNode := Doc.CreateTextNode('');
    ItemNode.AppendChild(TextNode);
    result := ItemNode;
  end;

end.


